source/utilities.tex

```diff
@@ -4442,6 +4451,551 @@
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+This type
+models \libconcept{contiguous_iterator}\iref{iterator.concept.contiguous},
+meets the \oldconcept{RandomAccessIterator} requirements\iref{random.access.iterators},
+and meets the requirements for constexpr iterators\iref{iterator.requirements.general},
+with value type \tcode{remove_cv_t<T>}.
+The reference type is \tcode{T\&} for \tcode{iterator}.
+
+\pnum
+All requirements on container iterators\iref{container.reqmts} apply to
+\tcode{optional::iterator}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr iterator begin() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+  If \tcode{has_value()} is \tcode{true},
+  an iterator referring to \tcode{*\exposid{val}}.
+  Otherwise, a past-the-end iterator value.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr iterator end() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{begin() + has_value()}.
+\end{itemdescr}
+
+\rSec3[optional.ref.observe]{Observers}
+
+\begin{itemdecl}
+constexpr T* operator->() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\hardexpects
+\tcode{has_value()} is \tcode{true}.
+
+\pnum
+\returns
+\exposid{val}.
+
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr T& operator*() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\hardexpects
+\tcode{has_value()} is \tcode{true}.
+
+\pnum
+\returns
+\tcode{*\exposid{val}}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr explicit operator bool() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{\exposid{val} != nullptr}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr bool has_value() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{\exposid{val} != nullptr}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr T& value() const;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+return has_value() ? *@\exposid{val}@ : throw bad_optional_access();
+\end{codeblock}
+\end{itemdescr}
+
+\begin{itemdecl}
+template<class U = remove_cv_t<T>> constexpr remove_cv_t<T> value_or(U&& u) const;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+ Let \tcode{X} be \tcode{remove_cv_t<T>}.
+
+\pnum
+\mandates
+\tcode{is_constructible_v<X, T\&> \&\& is_convertible_v<U, X>} is \tcode{true}.
+
+\pnum
+\effects
+ Equivalent to:
+\begin{codeblock}
+return has_value() ? *@\exposid{val}@ : static_cast<X>(std::forward<U>(u));
+\end{codeblock}
+\end{itemdescr}
+
+\rSec3[optional.ref.monadic]{Monadic operations}
+
+\begin{itemdecl}
+template<class F> constexpr auto and_then(F&& f) const;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+Let \tcode{U} be \tcode{invoke_result_t<F, T\&>}.
+
+\pnum
+\mandates
+\tcode{remove_cvref_t<U>} is a specialization of \tcode{optional}.
+
+\pnum
+\effects
+  Equivalent to:
+\begin{codeblock}
+if (has_value()) {
+  return invoke(std::forward<F>(f), *@\exposid{val}@);
+} else {
+  return remove_cvref_t<U>();
+}
```

Comment:
IIRC LWG didn't want these braces here, but it's consistent with the equivalent functions for `optional<T>`. We can remove them from both places later if necessary.

Reply:
Hm, surely this is editorial? Why is LWG opposed to braces? Have braces ever decreased clarity of presentation? goto fail; goto fail.